crates/kubelet/src/volumes.rs

```diff
@@ -0,0 +1,186 @@
+                    Ok((
+                        v.name.to_owned(),
+                        VolumeRef {
+                            host_path,
+                            volume_type,
+                        },
+                    ))
+                }
+            });
+            futures::future::join_all(volumes)
+                .await
+                .into_iter()
+                .collect()
+        } else {
+            Ok(HashMap::default())
+        }
+    }
+}
+
+impl AsRef<PathBuf> for VolumeRef {
+    fn as_ref(&self) -> &PathBuf {
+        &self.host_path
+    }
+}
+
+impl Deref for VolumeRef {
+    type Target = PathBuf;
+
+    fn deref(&self) -> &Self::Target {
+        &self.host_path
+    }
+}
+
+impl Drop for VolumeRef {
+    fn drop(&mut self) {
+        if matches!(self.volume_type, VolumeType::ConfigMap | VolumeType::Secret) {
+            // TODO: Currently there is no way to do this async (though there is an async destructors proposal)
+            debug!(
+                "deleting {:?} directory {:?}",
+                self.volume_type, self.host_path
+            );
+            std::fs::remove_dir_all(&self.host_path).unwrap_or_else(|e| {
+                error!(
+                    "unable to delete directory {:?} on volume cleanup: {:?}",
+                    self.host_path, e
+                )
+            });
+        }
+    }
+}
+
+/// This is a gnarly function to check all of the supported data members of the Volume struct.
+/// Because it isn't a HashMap, we need to check all fields individually
+async fn configure(
+    vol: &KubeVolume,
+    namespace: &str,
+    client: &kube::Client,
+    path: &PathBuf,
+) -> anyhow::Result<VolumeType> {
+    if let Some(cm) = &vol.config_map {
+        populate_from_config_map(
+            &cm.name
+                .as_ref()
+                .ok_or_else(|| anyhow::anyhow!("no configmap name was given"))?,
+            namespace,
+            client,
+            path,
+        )
+        .await
+    } else if let Some(s) = &vol.secret {
+        populate_from_secret(
+            &s.secret_name
+                .as_ref()
+                .ok_or_else(|| anyhow::anyhow!("no configmap name was given"))?,
+            namespace,
+            client,
+            path,
+        )
+        .await
+    } else {
+        Err(anyhow::anyhow!(
+            "Unsupported volume type. Currently supported types: ConfigMap, Secret, and HostPath"
+        ))
+    }
+}
+
+async fn populate_from_secret(
+    name: &str,
+    namespace: &str,
+    client: &kube::Client,
+    path: &PathBuf,
+) -> anyhow::Result<VolumeType> {
+    tokio::fs::create_dir(path).await?;
+    let secret_client: Api<Secret> = Api::namespaced(client.clone(), namespace);
+    let secret = secret_client.get(name).await?;
+    let data = secret.data.unwrap_or_default();
+    let data = data.iter().map(|(key, data)| async move {
```

Comment:
Can you deconstruct the `data` arg to the closure like this `ByteString(data)` so that you don't have to call `data.0`.

Reply:
Oooh right, I forgot about that handy thing